Runtime/GuidBasedReferenceList.cs

```diff
@@ -0,0 +1,166 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+#if UNITY_EDITOR
+using UnityEditor;
+#endif
+using UnityEngine;
+using Object = UnityEngine.Object;
+
+namespace RoyTheunissen.AssetPalette
+{
+    /// <summary>
+    /// There's two reasons why you should be using this class instead of having direct Object[] references:
+    /// 
+    /// - This way the actual loading of the assets is delayed as much as possible. The palette will not load any of
+    /// the references until it's necessary, for example for generating a preview or actually using a palette entry.
+    /// 
+    /// - This way the Personal Palette can correctly serialize and deserialize the asset reference. The Personal
+    /// Palette is not stored in an asset but is stored in the EditorPrefs as JSON instead. By default Unity's
+    /// JSON utility does not seem to deserialize asset references correctly and they will become corrupt.
+    /// </summary>
+    [Serializable]
+    public sealed class GuidBasedReferenceList<T> : IList<T> 
+        where T : Object
+    {
+        [SerializeField] private List<string> guids;
```

Comment:
Can't this be a `List<GuidBasedReference<T>>` so we don't have to duplicate a lot of the code, but just cast the final `List<T>`

Reply:
No I don't think so, in order to get a `T[]` out of that you need to do a lot of shenanigans and I don't want that to be in `PaletteEntry`.

If we expect all future PaletteEntries to use GUID based references it should be very easy to have a collection of them, as easy as just writing `T[]`